platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/maven/RealisedMavenModuleResolveMetadata.java

```diff
@@ -201,24 +202,26 @@ private static RealisedConfigurationMetadata createConfiguration(
     }
 
     static ImmutableList<? extends ModuleComponentArtifactMetadata> getArtifactsForConfiguration(DefaultMavenModuleResolveMetadata metadata) {
-        ImmutableList<? extends ModuleComponentArtifactMetadata> artifacts;
-
         if (metadata.isRelocated()) {
             // relocated packages have no artifacts
-            artifacts = ImmutableList.of();
+            return ImmutableList.of();
         } else if (metadata.isPomPackaging()) {
             // Modules with POM packaging _may_ have a jar
-            artifacts = ImmutableList.of(metadata.optionalArtifact("jar", "jar", null));
+            return ImmutableList.of(metadata.optionalArtifact("jar", "jar", null));
         } else if (metadata.isKnownJarPackaging()) {
             // Modules with a type of packaging that's always a jar
-            artifacts = ImmutableList.of(metadata.artifact("jar", "jar", null));
+            return ImmutableList.of(metadata.artifact("jar", "jar", null));
         } else {
-            // Modules with other types of packaging may publish an artifact with that extension or a jar
             String type = metadata.getPackaging();
-            artifacts = ImmutableList.of(new DefaultModuleComponentArtifactMetadata(metadata.getId(), new DefaultIvyArtifactName(metadata.getId().getModule(), type, type),
-                metadata.artifact("jar", "jar", null)));
+            // We were unable to resolve variable substitutions in the POM, so assume we're looking for a jar
+            if (PomReader.hasUnresolvedSubstitutions(type)) {
```

Comment:
Why is this static method located on the `PomReader` if it is not used there? It seems like it would be better located in this class

Reply:
I was trying to keep the knowledge about what a valid Maven substitution is somewhere close to where that substitution happens. The pattern used by hasUnresolvedSubstitutions is from IvyPatternHelper and that's extensively used by PomReader.

The "better" fix would be to bring IvyPatternHelper into gradle/gradle like we did with PomReader and move hasUnresolvedSubstitutions into that.